lib/jara/releaser.rb

```diff
@@ -28,30 +28,34 @@ def initialize(environment, bucket_name, options={})
 
     def build_artifact
       if @environment.nil?
-        jar_name = "#{app_name}.jar"
+        archive_name = "#{app_name}.#{@archiver.extension}"
         Dir.chdir(project_dir) do
-          @archiver.create(jar_name: jar_name)
+          @archiver.create(archive_name: archive_name)
         end
         @logger.info('Created test artifact')
-        File.join(project_dir, 'build', jar_name)
+        File.join(project_dir, 'build', archive_name)
       elsif (artifact_path = find_local_artifact)
         @logger.warn('An artifact for %s already exists: %s' % [branch_sha[0, 8], File.basename(artifact_path)])
         artifact_path
       else
         date_stamp = Time.now.utc.strftime('%Y%m%d%H%M%S')
         destination_dir = File.join(project_dir, 'build', @environment)
-        jar_name = [app_name, @environment, date_stamp, branch_sha[0, 8]].join('-') << '.jar'
+        archive_name = [app_name, @environment, date_stamp, branch_sha[0, 8]].join('-') << '.' << @archiver.extension
         Dir.mktmpdir do |path|
           @shell.exec('git archive --format=tar --prefix=%s/ %s | (cd %s/ && tar xf -)' % [File.basename(path), branch_sha, File.dirname(path)])
           Dir.chdir(path) do
             @logger.info('Checked out %s from branch %s' % [branch_sha[0, 8], @branch])
-            @archiver.create(jar_name: jar_name)
+            if @build_command
+              @logger.info('Running build command: %s' % @build_command)
+              @shell.exec(@build_command)
+            end
```

Comment:
As it is neither tool (tar, Puck) has much concept of building, they're both archiving tools. You point them to something and they bundle it up into a single file. Puck is a bit more intelligent than tar because it figures out where to find things that are not in the directory you point it at, and it doesn't put them into the archive exactly the way they are layed out in the directory, but you could argue that tar doesn't either if you have symlinks in your directory structure.

Before the build command feature Jara did this:

* Create a clean copy of the code
* Create an archive
* Upload the archive

The missing step there is that many times the clean copy of the code isn't enough, you also want to build something _from the code_. I see that as a separate step that needs to go between the two first: the build step.

It's separate because different build systems could use the same archiver (tar works both for plain JavaScript with minification and for CoffeScript, but they require different builds), and a build system could use different archivers depending on what you wanted (package this JRuby code with Java extensions as a JAR for Hadoop with Rubydoop or as a standalone application with Puck, for example).

There are of course cases when the build can't be separated from the archive step, but I can live with those being just an archiver and you'd just not use the build command feature. The build is optional.